wsagent/che-core-api-debugger/src/main/java/org/eclipse/che/api/debugger/server/DebuggerActionProvider.java

```diff
@@ -0,0 +1,100 @@
+/*******************************************************************************
+ * Copyright (c) 2012-2016 Codenvy, S.A.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.api.debugger.server;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+
+import org.eclipse.che.api.debugger.shared.dto.action.Action;
+import org.eclipse.che.api.debugger.shared.dto.action.ResumeAction;
+import org.eclipse.che.api.debugger.shared.dto.action.StartAction;
+import org.eclipse.che.api.debugger.shared.dto.action.StepIntoAction;
+import org.eclipse.che.api.debugger.shared.dto.action.StepOutAction;
+import org.eclipse.che.api.debugger.shared.dto.action.StepOverAction;
+import org.eclipse.che.dto.server.DtoFactory;
+import org.eclipse.che.dto.shared.DTO;
+
+import javax.inject.Singleton;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.MultivaluedMap;
+import javax.ws.rs.ext.MessageBodyReader;
+import javax.ws.rs.ext.Provider;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Type;
+
+/**
+ * Implementation of {@link MessageBodyReader} needed for binding JSON content to Java Objects.
+ *
+ * @author Anatoliy Bazko
+ * @see DTO
+ * @see DtoFactory
+ */
+@Singleton
+@Provider
+@Consumes({MediaType.APPLICATION_JSON})
+public class DebuggerActionProvider implements MessageBodyReader<Action> {
+
+    @Override
+    public boolean isReadable(Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType) {
+        return Action.class.isAssignableFrom(type);
+    }
+
+    @Override
+    public Action readFrom(Class<Action> type,
+                           Type genericType,
+                           Annotation[] annotations,
+                           MediaType mediaType,
+                           MultivaluedMap<String, String> httpHeaders,
+                           InputStream entityStream) throws IOException, WebApplicationException {
+        String json = readJson(entityStream);
```

Comment:
`JsonParser` can parse json from the reader, there is no need to read body as string.